pkg/common/cns-lib/volume/manager.go

```diff
@@ -184,161 +192,235 @@ func (m *defaultManager) ResetManager(ctx context.Context, vcenter *cnsvsphere.V
 	log.Infof("Done resetting volume.defaultManager")
 }
 
-// CreateVolume creates a new volume given its spec.
-func (m *defaultManager) CreateVolume(ctx context.Context, spec *cnstypes.CnsVolumeCreateSpec) (*CnsVolumeInfo, error) {
-	internalCreateVolume := func() (*CnsVolumeInfo, error) {
-		log := logger.GetLogger(ctx)
-		err := validateManager(ctx, m)
-		if err != nil {
-			return nil, err
+// createVolumeWithIdempotency leverages the VolumeOperationRequest interface to persist CNS task information. It uses
+// this persisted information to handle idempotency of CreateVolume callbacks to CNS for the same volume.
+func (m *defaultManager) createVolumeWithIdempotency(ctx context.Context, spec *cnstypes.CnsVolumeCreateSpec) (
+	*CnsVolumeInfo, error) {
+	log := logger.GetLogger(ctx)
+	var (
+		// Store the volume name passed in by input spec, this
+		// name may exceed 80 characters.
+		volNameFromInputSpec = spec.Name
+		// Reference to the CreateVolume task on CNS.
+		task *object.Task
+		// Local instance of CreateVolume details that needs to
+		// be persisted.
+		volumeOperationDetails *cnsvolumeoperationrequest.VolumeOperationRequestDetails
+	)
+
+	if m.operationStore == nil {
+		return nil, logger.LogNewError(log, "operation store cannot be nil")
+	}
+
+	// Determine if CNS CreateVolume needs to be invoked.
+	volumeOperationDetails, err := m.operationStore.GetRequestDetails(ctx, volNameFromInputSpec)
+	switch {
+	case err == nil:
+		if volumeOperationDetails.OperationDetails != nil {
+			// Validate if previous attempt was successful.
+			if volumeOperationDetails.OperationDetails.TaskStatus == taskInvocationStatusSuccess &&
+				volumeOperationDetails.VolumeID != "" {
+				log.Infof("Volume with name %s and id %s is already created on CNS with opId: %q.",
```

Comment:
Use %q for strings at all places for consistency

Reply:
done